jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java

```diff
@@ -0,0 +1,376 @@
+import javax.servlet.http.HttpServletResponse;
+
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.jetty.http.MimeTypes;
+import org.eclipse.jetty.http.pathmap.PathSpecSet;
+import org.eclipse.jetty.server.HttpChannel;
+import org.eclipse.jetty.server.HttpOutput;
+import org.eclipse.jetty.server.HttpOutput.Interceptor;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.IncludeExclude;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Buffered Response Handler
+ * <p>
+ * A Handler that can apply a {@link Interceptor}
+ * mechanism to buffer the entire response content until the output is closed.
+ * This allows the commit to be delayed until the response is complete and thus
+ * headers and response status can be changed while writing the body.
+ * <p>
+ * Note that the decision to buffer is influenced by the headers and status at the
+ * first write, and thus subsequent changes to those headers will not influence the
+ * decision to buffer or not.
+ * <p>
+ * Note also that there are no memory limits to the size of the buffer, thus
+ * this handler can represent an unbounded memory commitment if the content
+ * generated can also be unbounded.
+ * </p>
+ */
+public class FileBufferedResponseHandler extends HandlerWrapper
+{
+    static final Logger LOG = LoggerFactory.getLogger(FileBufferedResponseHandler.class);
+
+    private final IncludeExclude<String> _methods = new IncludeExclude<>();
+    private final IncludeExclude<String> _paths = new IncludeExclude<>(PathSpecSet.class);
+    private final IncludeExclude<String> _mimeTypes = new IncludeExclude<>();
+    private Path tempDir = new File(System.getProperty("java.io.tmpdir")).toPath();
+
+    public FileBufferedResponseHandler()
+    {
+        // include only GET requests
+
+        _methods.include(HttpMethod.GET.asString());
+        // Exclude images, aduio and video from buffering
+        for (String type : MimeTypes.getKnownMimeTypes())
+        {
+            if (type.startsWith("image/") ||
+                type.startsWith("audio/") ||
+                type.startsWith("video/"))
+                _mimeTypes.exclude(type);
+        }
+        LOG.debug("{} mime types {}", this, _mimeTypes);
+    }
+
+    public IncludeExclude<String> getMethodIncludeExclude()
+    {
+        return _methods;
+    }
+
+    public IncludeExclude<String> getPathIncludeExclude()
+    {
+        return _paths;
+    }
+
+    public IncludeExclude<String> getMimeIncludeExclude()
+    {
+        return _mimeTypes;
+    }
+
+    @Override
+    public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
```

Comment:
What happens if a large static file is served from DefaultServlet?
I would hope this Handler and Interceptor does not run for that scenario.

Reply:
I don't think that is always problem as the one and only one write from the ResourceService would come with last==true and thus aggregation would not be done.
But there are cases in the ResourceService where a large content is cycled through a buffer, so we don't want those aggregated.

Two options: detect cases that we don't want to aggregate, which may be a little complex but doable;  or document that the Path matching should be used to exclude URIs and/or mime types that this is likely to happen for (or only apply to URIs that this handling is needed for).